common/unsafe/src/main/java/org/apache/spark/unsafe/types/VariantVal.java

```diff
@@ -0,0 +1,124 @@
+ * and {@link VariantVal#metadata}. The value encodes types and values, but not field names. The
+ * metadata currently contains a version flag and a list of field names. We can extend/modify the
+ * detailed binary format given the version flag.
+ * <p>
+ * A {@link VariantVal} can be produced by casting another value into the Variant type or parsing a
+ * JSON string in the {@link org.apache.spark.sql.catalyst.expressions.variant.ParseJson}
+ * expression. We can extract a path consisting of field names and array indices from it, cast it
+ * into a concrete data type, or rebuild a JSON string from it.
+ * <p>
+ * The storage layout of this class in {@link org.apache.spark.sql.catalyst.expressions.UnsafeRow}
+ * and {@link org.apache.spark.sql.catalyst.expressions.UnsafeArrayData} is: the fixed-size part is
+ * a long value "offsetAndSize". The upper 32 bits is the offset that points to the start position
+ * of the actual binary content. The lower 32 bits is the total length of the binary content. The
+ * binary content contains: 4 bytes representing the length of {@link VariantVal#value}, content of
+ * {@link VariantVal#value}, content of {@link VariantVal#metadata}. This is an internal and
+ * transient format and can be modified at any time.
+ */
+public class VariantVal implements Serializable {
+  protected final byte[] value;
+  protected final byte[] metadata;
+
+  public VariantVal(byte[] value, byte[] metadata) {
+    this.value = value;
+    this.metadata = metadata;
+  }
+
+  public byte[] getValue() {
+    return value;
+  }
+
+  public byte[] getMetadata() {
+    return metadata;
+  }
+
+  /**
+   * This function writes the binary content into {@code buffer} starting from {@code cursor}, as
+   * described in the class comment. The caller should guarantee there is enough space in `buffer`.
+   */
+  public void writeIntoUnsafeRow(byte[] buffer, long cursor) {
+    Platform.putInt(buffer, cursor, value.length);
+    Platform.copyMemory(value, Platform.BYTE_ARRAY_OFFSET, buffer, cursor + 4, value.length);
+    Platform.copyMemory(
+        metadata,
+        Platform.BYTE_ARRAY_OFFSET,
+        buffer,
+        cursor + 4 + value.length,
+        metadata.length
+    );
+  }
+
+  /**
+   * This function reads the binary content described in `writeIntoUnsafeRow` from `baseObject`. The
+   * offset is computed by adding the offset in {@code offsetAndSize} and {@code baseOffset}.
+   */
+  public static VariantVal readFromUnsafeRow(long offsetAndSize, Object baseObject,
+                                             long baseOffset) {
+    // offset and totalSize is the upper/lower 32 bits in offsetAndSize.
+    int offset = (int) (offsetAndSize >> 32);
+    int totalSize = (int) offsetAndSize;
+    int valueSize = Platform.getInt(baseObject, baseOffset + offset);
+    int metadataSize = totalSize - 4 - valueSize;
+    byte[] value = new byte[valueSize];
+    byte[] metadata = new byte[metadataSize];
+    Platform.copyMemory(
```

Comment:
We can avoid copy if `VariantVal` can follow `UTF8String` and also represent its data as `baseObject + baseOffset`. Does Java have nicer way to do it now? cc @rednaxelafx

Reply:
I know how `UTF8String` works, but I feel it is simpler to have `byte[]` in the `VariantVal` object instead of `baseObject + baseOffset`. I prefer to have this version first and it is not something unchangeable in the future.